wgengine/wglog/wglog.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package wglog contains logging helpers for wireguard-go.
+package wglog
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"sync/atomic"
+
+	"github.com/tailscale/wireguard-go/device"
+	"github.com/tailscale/wireguard-go/wgcfg"
+	"tailscale.com/types/logger"
+)
+
+// A Logger is a wireguard-go log wrapper that cleans up and rewrites log lines.
+// It can be modified at run time to adjust to new wireguard-go configurations.
+type Logger struct {
+	DeviceLogger *device.Logger
+	replacer     atomic.Value // of *strings.Replacer
+}
+
+// NewLogger creates a new logger for use with wireguard-go.
+// This logger silences repetitive/unhelpful noisy log lines
+// and rewrites peer keys from wireguard-go into Tailscale format.
+func NewLogger(logf logger.Logf) *Logger {
+	ret := new(Logger)
+
+	wrapper := func(format string, args ...interface{}) {
+		msg := fmt.Sprintf(format, args...)
+		if strings.Contains(msg, "Routine:") {
+			// wireguard-go logs as it starts and stops routines.
+			// Drop those; there are a lot of them, and they're just noise.
+			return
+		}
+		r := ret.replacer.Load()
+		if r == nil {
+			// No replacements specified; log as originally planned.
+			logf(format, args...)
+			return
+		}
+		// Do the replacements.
+		new := r.(*strings.Replacer).Replace(msg)
+		if new == msg {
+			// No replacements. Log as originally planned.
+			logf(format, args...)
+			return
+		}
+		// We made some replacements. Log the new version.
+		// This changes the format string,
+		// which is somewhat unfortunate as it impacts rate limiting,
+		// but there's not much we can do about that.
+		logf("%s", new)
```

Comment:
I think this will be a problem. The rate-limiting at lower levels will rate-limit on this `"%s"` pattern as a whole.

Reply:
... which you said in the comment right above! But I think we need to deal with it.

Reply:
we know the set of log lines that wireguard produces. 

can we clean the args instead?

Reply:
As we discussed privately earlier, logger.StdLogger has already flattened the format to `%s` by the time it reaches here, so we have two things to fix: StdLogger and this line.

I tried implementing a few things here and am not happy with what I've found.

Starting with StdLogger, options:

(1) Have each StdLogger/FuncWriter do its own (non-format-string-based) rate limiting, and then add a non-rate-limited prefix to its output. This has a downside, which is that we don't ever strip off the "don't rate limit this" prefix, so there's extra junk in all the logs. Even if we do strip it off in the rate-limiter, the tests don't use the rate limiter, so you need a separate way to strip it off. It gets ugly. Plus you end up wanting to configure the StdLogger's rate-limiting params. Plus layering means that prefixes don't always stay prefixes, but sometimes end up in the middle of the string, which makes rate-limiting and subsequent cleanup more annoying and expensive.

(2) We could get rid of StdLogger in this particular case by changing wireguard-go to accept a Print{f,ln} interface, but we use StdLogger in a few other places as well. And accepting a Println significantly complicates this logic as well; we need to concoct a format string for Println, perhaps based on types. (Note that we can't detect string literals vs non-literals at run time.)


Assuming we solved the StdLogger problem, options for this line:

(1) We could add a special "wireguard-go" prefix to these lines and then strip them off in the rate limiter, but that has all the same problems as option (1) above.

(2) Clean each arg instead. For correctness, we need to parse the format string to check the verb used with each argument. This is complicated and expensive. Or we need to write a tool to check that no unusual (non-%v) format strings are in use anywhere and add that CI. Same for checking whether types satisfy fmt.Stringer. We either recreate large parts of package fmt or write tools to ensure that we don't need to or hope that we'll notice if something breaks in the future.


I'm coming to the conclusion that this is all a pretty unpleasant limitation of the `logger.Logf` + format-string-based rate-limiting approach.

As a general workaround to the limitations of this approach, we could make magic arg types that signal information to `logger.Logf`s; you would add extra args to inform other layers things like "use this format string for ratelimiting" and "I've already rate-limited this" and "verbosity level 1". The downside to these arg types is that anyone who naively did a `fmt.Sprintf(format, args...)` would end up with failures, because of the extra args, so we'd need some helper functions to filter out those args before actually doing the formatting.

That's the most appealing approach I've come up with so far, but it's a significant design change and it warrants discussion.

---

All that said, I propose that we go ahead with this PR as written, on the grounds that it is no worse than what we already had. We already used a StdLogger for wireguard-go, so we already had these exact rate-limiting problems in this exact same place. Yes, there is an outstanding issue about that. But I don't see a clear path forward, and certainly not a clear path forward that I can implement in the negative forty-five minutes before the end of my day.